src/swap/components/SwapProvider.tsx

```diff
@@ -184,21 +189,85 @@ export function SwapProvider({
           return handleError({ swapError: response });
         }
 
-        onSubmit?.(response);
+        const { transaction, approveTransaction } = response;
+
+        /* for swaps from ERC-20 tokens, 
+          if there is an approveTransaction present,
+          request approval for the amount  */
+        if (approveTransaction?.data) {
+          setPendingTransaction(true);
+          const approveTxHash = await sendTransactionAsync({
+            to: approveTransaction.to,
+            value: approveTransaction.value,
+            data: approveTransaction.data,
+          });
+          await waitForTransactionReceipt(config, {
+            hash: approveTxHash,
+            confirmations: 1,
+          });
+          setPendingTransaction(false);
+        }
+
```

Comment:
There is a lot of code that miss unit-test in this file. And this code is super critical.
Let's take the time to unpack this Provider in utility functions that can be atomically tested.